dungeon-crawler/app/src/main/java/com/novoda/dungeoncrawler/JoystickView.java

```diff
@@ -0,0 +1,79 @@
+package com.novoda.dungeoncrawler;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.AttributeSet;
+import android.view.LayoutInflater;
+import android.view.MotionEvent;
+import android.view.View;
+import android.widget.FrameLayout;
+
+public class JoystickView extends FrameLayout {
+
+    private enum Action {
+        NONE, BACKWARD, ATTACK, FORWARD
+    }
+
+    private static final Handler MAIN_THREAD = new Handler(Looper.getMainLooper());
+    private static final long ACTION_ATTACK_DELAY_MS = 1000;
+    private static final long ACTION_MOVE_DELAY_MS = 20;
+
+    private Action userAction = Action.NONE;
+
+    public JoystickView(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        LayoutInflater.from(context).inflate(R.layout.joystick, this, true);
+    }
+
+    @Override
+    protected void onFinishInflate() {
+        super.onFinishInflate();
+        findViewById(R.id.backward).setOnTouchListener(this::onBackward);
+        findViewById(R.id.attack).setOnClickListener(this::onAttack);
+        findViewById(R.id.forward).setOnTouchListener(this::onForward);
+    }
+
+    private void onAttack(View _) {
+        userAction = Action.ATTACK;
+        delayResetAction(ACTION_ATTACK_DELAY_MS);
+    }
+
+    private boolean onBackward(View _, MotionEvent motionEvent) {
+        updateActionWith(motionEvent, Action.BACKWARD);
+        return true;
+    }
+
+    private boolean onForward(View _, MotionEvent motionEvent) {
+        updateActionWith(motionEvent, Action.FORWARD);
+        return true;
+    }
+
+    private void delayResetAction(long actionResetDelayMs) {
+        MAIN_THREAD.removeCallbacks(resetLastAction);
+        MAIN_THREAD.postDelayed(resetLastAction, actionResetDelayMs);
+    }
+
+    private final Runnable resetLastAction = () -> userAction = Action.NONE;
+
+    private void updateActionWith(MotionEvent motionEvent, Action thisAction) {
+        int action = motionEvent.getActionMasked();
+        if (action == MotionEvent.ACTION_DOWN) {
+            userAction = thisAction;
+        } else if (action == MotionEvent.ACTION_UP) {
+            delayResetAction(ACTION_MOVE_DELAY_MS);
```

Comment:
don't need this because the game polls for the `tilt` state in the joystick impl